src/components/Document/DocumentText.vue

```diff
@@ -16,122 +22,73 @@ import { mapGetters } from 'vuex';
 
 export default {
 	computed: {
-		compiledHighlight() {
-			return {
-				render: h => {
-					return h(
-						Vue.compile(
-							`<p class="text-base md:text-lg">${this.highlightedText()}</p>`
-						)
-					);
-				}
-			};
+		documentSubstrings() {
+			return this.createDocumentSubstrings();
 		},
-		...mapGetters('DocumentStore', ['document', 'matches'])
+		...mapGetters('DocumentStore', ['document', 'matches', 'intervals'])
 	},
 	methods: {
-		highlightedText: function () {
-			const indices = this.matches
-				.map(matched_doc =>
-					matched_doc.intervals.map(match => ({
-						name: matched_doc.name,
-						from: match.begin,
-						to: match.end,
-						percentage: this.$filters.roundToTwoDecimals(
-							this.$filters.toNumber(matched_doc.percentage)
-						),
-						color: colorForIndex(
-							this.matches
-								.map(d => d.name)
-								.indexOf(matched_doc.name)
-						)
-					}))
-				)
-				.flat();
-
-			const intervals = this.mergeIntervals(indices);
-
-			const subStringsToReplace = intervals.map(h => ({
-				text: this.document.text.substring(h.from, h.to + 1),
-				docs: h.matches,
-				color: h.color
-			}));
+		createDocumentSubstrings: function () {
+			
+			const documentIds = this.intervals.flatMap(({ matches }) =>
+				matches.map(m => m.id)
+			);
 
-			return subStringsToReplace.reduce(
-				(string, substring) =>
-					string.replace(
-						new RegExp(substring.text, 'g'),
-						`<span v-tooltip.top="'${this.toggleTooltip(
-							substring.docs
-						)}'" class="font-bold" style="color:${
-							substring.color
-						};">${substring.text}</span>`
-					),
-				escape(this.document.text)
+			const uniqueDocumentIds = documentIds.filter(
+				(x, i, a) => a.indexOf(x) == i
 			);
-		},
-		mergeIntervals: function (intervals) {
-			if (intervals.length <= 1) {
-				return intervals;
-			}
 
-			let stack = [];
-			let top = null;
+			const colors = uniqueDocumentIds.reduce((colors, documentId) => {
+				colors[documentId] = colorForIndex(documentId);
+				return colors;
+			}, {});
 
-			intervals = intervals.sort(function (a, b) {
-				if (a.from === b.from) {
-					if (a.to > b.to || a.to === b.to) {
-						return 1;
-					} else {
-						return -1;
-					}
-				} else {
-					if (a.from < b.from) {
-						return -1;
-					} else if (a.from > b.from) {
-						return 1;
-					}
-					return 0;
-				}
+			this.intervals.forEach(interval => {
+				const matches = interval.matches.slice();
+				const maxPercentageDocId = matches.sort((a, b) =>
+					a.percentage > b.percentage ? -1 : 1
+				)[0].id;
+				interval.color = colors[maxPercentageDocId];
 			});
 
-			intervals[0].matches = [
-				{ doc: intervals[0].name, per: intervals[0].percentage }
-			];
-			stack.push(intervals[0]);
+			let start = 0;
+			let textSubstrings = [];
 
-			intervals.shift();
-
-			intervals.forEach(item => {
-				top = stack[stack.length - 1];
+			this.intervals.forEach(interval => {
```

Comment:
```javascript
const textSubstrings = intervals.reduce(
  (substrings, currentRange, index) => {
    
    // previous range
    const previous = substrings[substrings.length - 1];
    previous.to = currentRange.from;
    previous.text = text.substring(previous.from,previous.to);
    
    // current
    currentRange.text = text.substring(currentRange.from,currentRange.to);
    substrings.push(currentRange);
    
    // following range (set starting index)
    const following = { from: currentRange.to + 1 };

    // in case of last element, copy the rest of the text till the end
    if (index === intervals.length -1) {
      following.to = text.length;
      following.text = text.substring(following.from,following.to)
    }
    substrings.push(following);

    return substrings;
  },
  [{ from: 0 }] // initial value - its `to` will be filled in first iteration.
);


return textSubstrings;
```